packages/qwik-city/src/middleware/request-handler/user-response.ts

```diff
@@ -50,62 +52,85 @@ export function runQwikCity<T>(
     qwikSerializer,
     resolve!
   );
+
   return {
     response: responsePromise,
     requestEv,
     completion: asyncStore
-      ? asyncStore.run(requestEv, runNext, requestEv, resolve!)
-      : runNext(requestEv, resolve!),
+      ? asyncStore.run(requestEv, runNext, requestEv, applyRewrite, resolve!)
+      : runNext(requestEv, applyRewrite, resolve!),
   };
 }
 
-async function runNext(requestEv: RequestEventInternal, resolve: (value: any) => void) {
-  try {
-    // Run all middlewares
-    await requestEv.next();
-  } catch (e) {
-    if (e instanceof RedirectMessage) {
-      const stream = requestEv.getWritableStream();
-      await stream.close();
-    } else if (e instanceof ServerError) {
-      if (!requestEv.headersSent) {
-        const status = e.status as StatusCodes;
-        const accept = requestEv.request.headers.get('Accept');
-        if (accept && !accept.includes('text/html')) {
-          const qwikSerializer = requestEv[RequestEvQwikSerializer];
-          requestEv.headers.set('Content-Type', 'application/qwik-json');
-          requestEv.send(status, await qwikSerializer._serializeData(e.data, true));
-        } else {
-          const html = getErrorHtml(e.status, e.data);
-          requestEv.html(status, html);
+async function runNext(
+  requestEv: RequestEventInternal,
+  applyRewrite: ApplyRewriteInternal,
+  resolve: (value: any) => void
+) {
+  async function _runNext(
```

Comment:
@wmertens If you have any suggestions as to how to make this function clearer, I'm all ears. It became more complex after introduced with the rewrite handling.